src/libPMacc/include/traits/GetUniqueTypeId.hpp

```diff
@@ -57,27 +60,47 @@ const uint64_t GetUniqueTypeId<T_Type>::id = uint64_t(&GetUniqueTypeId<T_Type>::
 
 /** Get a unique id of a type
  *
- * - generate a unique 64bit id of a type at *runtime*
+ * - generate a unique id of a type at *runtime*
  * - the id of a type is equal on each instance of a process
  *
  * @tparam T_Type any object (class or typename)
- *
- * @treturn ::uid
+ * @tparam T_ResultType result type
  */
-template<typename T_Type>
+template<typename T_Type, typename T_ResultType = uint64_t>
 struct GetUniqueTypeId
 {
-    static const uint64_t uid;
-};
 
-/** instantiation of traits::GetUniqueTypeId
- *
- * - create a instance of `traits::GetUniqueTypeId` and initialize the uid
- * - map `detail::GetUniqueTypeId<T>` to a range [1; 2^64-1]
- *   `uid = unique_id_of_current_type - base_unique_id`
- */
-template<typename T_Type>
-const uint64_t GetUniqueTypeId<T_Type>::uid = detail::GetUniqueTypeId<T_Type>::id - detail::GetUniqueTypeId<uint8_t>::id;
+    /** create unique id
+     *
+     * @param maxValue largest allowed id
+     */
+    static const T_ResultType uid(uint64_t maxValue = uint64_t(1) << (sizeof (T_ResultType) * 8) - uint64_t(1))
+    {
+        typedef T_ResultType ResultType;
```

Comment:
Why do you typedef ```ResultType``` from ```T_ResultType``` but directly use ```T_Type``` within the method?
Nevertheless, I can see a usage for both of them being defined at class scope.